File: notify.go
// Real OS notifications — the kind WhatsApp raises, with collab's own name and
// icon on them.
//
// Neither platform will let a plain Go binary do this. macOS attributes a
// notification to an application bundle, so the Mac side is a small signed
// Swift .app (notify/mac) that this shells out to. Windows attributes a toast
// to a registered AppUserModelID, so the Windows side is a small C# helper
// (notify/windows). Both are found next to this binary, and if neither is
// there collab simply stays quiet rather than falling back to something that
// pops up under another app's name.
package main

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

// findNotifier returns the helper to run, or "" if this machine has none.
func findNotifier() string {
	var candidates []string
	exe, err := os.Executable()
	if err == nil {
		if p, err := filepath.EvalSymlinks(exe); err == nil {
			exe = p
		}
	}
	dir := filepath.Dir(exe)

	switch runtime.GOOS {
	case "darwin":
		for _, base := range []string{dir, home("Applications"), "/Applications"} {
			candidates = append(candidates, filepath.Join(base, "collab.app", "Contents", "MacOS", "collab-notify"))
		}
	case "windows":
		candidates = append(candidates,
			filepath.Join(dir, "collab-notify.exe"),
			filepath.Join(dir, "notify", "collab-notify.exe"))
	}
	for _, c := range candidates {
		if st, err := os.Stat(c); err == nil && !st.IsDir() {
			return c
		}
	}
	return ""
}

func notifyEnabled() bool { return env("COLLAB_NOTIFY", "1") != "0" }

// raise posts one notification. Failures are reported once and then ignored —
// a broken notifier must not take the watcher down with it.
var notifyWarned bool

func raise(helper, title, subtitle, body string) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()
	out, err := exec.CommandContext(ctx, helper, title, body, subtitle).CombinedOutput()
	if err != nil && !notifyWarned {
		notifyWarned = true
		fmt.Fprintf(os.Stderr, "* notifications are not working (%v) %s\n", err, strings.TrimSpace(string(out)))
	}
}

func ellipsis(s string, n int) string {
	r := []rune(s)
	if len(r) <= n {
		return s
	}
	return strings.TrimSpace(string(r[:n-1])) + "…"
}

// ─────────────────────── coalescing ───────────────────────
//
// Messages are not popped one at a time. When a machine wakes after being
// asleep the server replays everything it missed at once, and forty popups in
// a row is not a notification, it is a punishment. So arrivals are collected
// until the channel goes quiet, and a burst becomes one summary.

const notifyQuiet = 700 * time.Millisecond

type notifier struct {
	in     chan Msg
	helper string
	me     string
}

func newNotifier(me string) *notifier {
	helper := findNotifier()
	if helper == "" || !notifyEnabled() {
		return nil
	}
	n := &notifier{in: make(chan Msg, 512), helper: helper, me: me}
	go n.run()
	return n
}

func (n *notifier) send(m Msg) {
	if n == nil {
		return
	}
	select {
	case n.in <- m:
	default: // never block the watcher for the sake of a popup
	}
}

func (n *notifier) run() {
	var pending []Msg
	var timer *time.Timer
	var fire <-chan time.Time
	for {
		select {
		case m := <-n.in:
			if strings.EqualFold(strings.TrimSpace(m.From), strings.TrimSpace(n.me)) {
				continue // your own words do not need announcing back to you
			}
			pending = append(pending, m)
			if timer == nil {
				timer = time.NewTimer(notifyQuiet)
			} else {
				if !timer.Stop() {
					select {
					case <-timer.C:
					default:
					}
				}
				timer.Reset(notifyQuiet)
			}
			fire = timer.C

		case <-fire:
			n.flush(pending)
			pending, fire = nil, nil
		}
	}
}

func (n *notifier) flush(ms []Msg) {
	switch len(ms) {
	case 0:
		return
	case 1:
		m := ms[0]
		if m.kind() == KindChange {
			raise(n.helper, m.From, fmt.Sprintf("%s · %s", m.Action, m.Target), ellipsis(m.Text, 180))
			return
		}
		raise(n.helper, m.From, "#"+m.Channel, ellipsis(m.Text, 180))
		return
	}

	var senders []string
	seen := map[string]bool{}
	changes := 0
	for _, m := range ms {
		if !seen[m.From] {
			seen[m.From] = true
			senders = append(senders, m.From)
		}
		if m.kind() == KindChange {
			changes++
		}
	}

	title := strings.Join(senders, " & ")
	if len(senders) > 2 {
		title = "collab"
	}
	sub := fmt.Sprintf("%d new on #%s", len(ms), ms[len(ms)-1].Channel)
	if changes > 0 {
		sub = fmt.Sprintf("%d new on #%s · %d change%s", len(ms), ms[len(ms)-1].Channel, changes, plural(changes))
	}
	last := ms[len(ms)-1]
	raise(n.helper, title, sub, ellipsis(last.From+": "+last.line(), 180))
}

func plural(n int) string {
	if n == 1 {
		return ""
	}
	return "s"
}

// testNotify is `collab test-notify` — the way to check the popups work
// without waiting for the other person to say something.
func testNotify() {
	h := findNotifier()
	if h == "" {
		fmt.Fprintf(os.Stderr, "collab: no notifier installed for %s\n", runtime.GOOS)
		switch runtime.GOOS {
		case "darwin":
			fmt.Fprintln(os.Stderr, "        put collab.app next to the collab binary, or in ~/Applications")
		case "windows":
			fmt.Fprintln(os.Stderr, "        put collab-notify.exe next to collab.exe")
		}
		os.Exit(1)
	}
	fmt.Printf("using %s\n", h)
	raise(h, "collab", "#"+channel()+" · test", "If you can see this, notifications work.")
	if !notifyWarned {
		fmt.Println("sent — it should be on screen now (and in Notification Centre)")
	}
}

File: core/src/notify.rs
//! Real OS notifications, raised by the platform helper.
//!
//! Neither platform lets a plain binary do this: macOS attributes a
//! notification to an application bundle and Windows to a registered
//! AppUserModelID, so the popup comes from Collab.app or collab-notify.exe. If
//! neither is present collab stays quiet rather than falling back to something
//! that pops up under another app's name.
use crate::config;
use crate::msg::Msg;
use std::path::PathBuf;
use std::process::Command;
#[cfg(not(target_os = "macos"))]
use std::sync::mpsc::{sync_channel, RecvTimeoutError};
use std::sync::mpsc::SyncSender;
use std::time::Duration;

/// A machine waking from sleep is replayed everything it missed at once, and
/// forty popups in a row is not a notification, it is a punishment. Arrivals
/// are collected until the channel goes quiet.
#[cfg(not(target_os = "macos"))]
const QUIET: Duration = Duration::from_millis(700);

/// On macOS this is Collab.app, which raises notifications itself: macOS
/// attributes a notification to a bundle's main executable, so a second binary
/// inside the same bundle is refused outright — and this binary has no bundle
/// at all. On Windows it is the standalone toast helper.
pub fn find_notifier() -> Option<PathBuf> {
    let exe = std::env::current_exe().ok()?;
    let exe = std::fs::canonicalize(&exe).unwrap_or(exe);
    let dir = exe.parent()?.to_path_buf();

    let mut candidates: Vec<PathBuf> = Vec::new();
    if cfg!(target_os = "macos") {
        for base in [
            dir.clone(),
            config::home("Applications"),
            PathBuf::from("/Applications"),
        ] {
            candidates.push(base.join("Collab.app"));
        }
    } else if cfg!(target_os = "windows") {
        candidates.push(dir.join("collab-notify.exe"));
        candidates.push(dir.join("notify/collab-notify.exe"));
    }
    candidates.into_iter().find(|p| p.exists())
}

/// Hands a request to the app — the only thing here allowed to raise one.
/// `-g` keeps it from stealing focus.
#[cfg(target_os = "macos")]
fn ask_app(url: &str) -> std::io::Result<()> {
    match Command::new("open").args(["-g", url]).status() {
        Ok(s) if s.success() => Ok(()),
        Ok(_) => Err(std::io::Error::other("Collab.app did not answer")),
        Err(e) => Err(e),
    }
}

#[cfg(target_os = "macos")]
pub fn ensure_app_running() -> bool {
    let Some(app) = find_notifier() else { return false };
    Command::new("open")
        .args(["-g", "-a"])
        .arg(&app)
        .status()
        .map(|s| s.success())
        .unwrap_or(false)
}

#[cfg(not(target_os = "macos"))]
fn gui_url() -> String {
    format!(
        "http://127.0.0.1:{}",
        config::env("COLLAB_GUI_PORT", "8788")
    )
}

#[cfg(not(target_os = "macos"))]
fn ellipsis(s: &str, n: usize) -> String {
    let chars: Vec<char> = s.chars().collect();
    if chars.len() <= n {
        return s.to_string();
    }
    format!("{}…", chars[..n - 1].iter().collect::<String>().trim_end())
}

pub struct Notifier {
    tx: SyncSender<Msg>,
}

impl Notifier {
    pub fn new(me: String) -> Option<Notifier> {
        let helper = find_notifier()?;
        if !config::notify_enabled() {
            return None;
        }

        // On macOS the app watches the channel itself and posts its own
        // notifications, so the right thing here is to make sure it is running
        // and then stay out of the way — two notifiers would double every popup.
        #[cfg(target_os = "macos")]
        {
            let _ = (me, helper);
            ensure_app_running();
            None
        }

        #[cfg(not(target_os = "macos"))]
        {
            let (tx, rx) = sync_channel::<Msg>(512);
        std::thread::spawn(move || {
            let mut pending: Vec<Msg> = Vec::new();
            let mut warned = false;
            loop {
                let wait = if pending.is_empty() {
                    Duration::from_secs(3600)
                } else {
                    QUIET
                };
                match rx.recv_timeout(wait) {
                    Ok(m) => {
                        // Your own words do not need announcing back to you —
                        // but your own AI's do. A name belongs to a machine, so
                        // without the second test this would also silence the
                        // assistant sitting next to you.
                        if m.via != crate::msg::ACTOR_AI
                            && m.from.trim().eq_ignore_ascii_case(me.trim())
                        {
                            continue;
                        }
                        // Addressed to somebody else: still delivered, not announced.
                        if !m.is_for(&config::my_names()) {
                            continue;
                        }
                        pending.push(m);
                    }
                    Err(RecvTimeoutError::Timeout) => {
                        flush(&helper, &pending, &mut warned);
                        pending.clear();
                    }
                    Err(RecvTimeoutError::Disconnected) => return,
                }
            }
        });
            Some(Notifier { tx })
        }
    }

    pub fn send(&self, m: &Msg) {
        let _ = self.tx.try_send(m.clone()); // never block the watcher for a popup
    }
}

#[cfg(not(target_os = "macos"))]
fn raise(
    helper: &PathBuf,
    title: &str,
    subtitle: &str,
    body: &str,
    channel: &str,
    warned: &mut bool,
) {
    // The last arguments are what a click needs: where the window is, what to
    // run to open it, and which channel the message came from.
    let self_path = std::env::current_exe().unwrap_or_default();
    let out = Command::new(helper)
        .args([
            title,
            body,
            subtitle,
            &gui_url(),
            &self_path.to_string_lossy(),
            channel,
            &config::name(),
        ])
        .output();
    match out {
        Ok(o) if !o.status.success() && !*warned => {
            *warned = true;
            eprintln!(
                "* notifications are not working ({}) {}",
                o.status,
                String::from_utf8_lossy(&o.stderr).trim()
            );
        }
        Err(e) if !*warned => {
            *warned = true;
            eprintln!("* notifications are not working ({e})");
        }
        _ => {}
    }
}

#[cfg(not(target_os = "macos"))]
fn flush(helper: &PathBuf, ms: &[Msg], warned: &mut bool) {
    match ms.len() {
        0 => {}
        1 => {
            let m = &ms[0];
            // The machine rides along in the title: a chat that named itself
            // "shop" says nothing about whose Claude it is, and that is the
            // question the popup has to answer.
            if m.is_change() {
                raise(
                    helper,
                    &m.label(),
                    &format!("{} · {}", m.action, m.target),
                    &ellipsis(&m.text, 180),
                    &m.channel,
                    warned,
                );
            } else {
                raise(
                    helper,
                    &m.label(),
                    &format!("#{}", m.channel),
                    &ellipsis(&m.text, 180),
                    &m.channel,
                    warned,
                );
            }
        }
        n => {
            let mut senders: Vec<String> = Vec::new();
            let mut changes = 0;
            for m in ms {
                if !senders.contains(&m.label()) {
                    senders.push(m.label());
                }
                if m.is_change() {
                    changes += 1;
                }
            }
            let title = if senders.len() > 2 {
                "collab".to_string()
            } else {
                senders.join(" & ")
            };
            let last = &ms[ms.len() - 1];
            let sub = if changes > 0 {
                format!(
                    "{n} new on #{} · {changes} change{}",
                    last.channel,
                    if changes == 1 { "" } else { "s" }
                )
            } else {
                format!("{n} new on #{}", last.channel)
            };
            let body = ellipsis(&format!("{}: {}", last.label(), last.line()), 180);
            raise(helper, &title, &sub, &body, &last.channel, warned);
        }
    }
}

/// macOS: the app is what posts notifications, so ask it.
#[cfg(target_os = "macos")]
pub fn test_notify() {
    let Some(app) = find_notifier() else {
        eprintln!("collab: Collab.app is not installed — run ./install.sh");
        std::process::exit(1);
    };
    println!("asking {} to post one", app.display());
    if !ensure_app_running() {
        eprintln!("collab: could not start Collab.app");
        std::process::exit(1);
    }
    std::thread::sleep(Duration::from_millis(700)); // let it finish launching
    match ask_app("collab://test") {
        Ok(()) => println!("sent — it should be on screen now (and in Notification Centre)"),
        Err(e) => {
            eprintln!("collab: {e}");
            std::process::exit(1);
        }
    }
}

#[cfg(not(target_os = "macos"))]
pub fn test_notify() {
    let Some(h) = find_notifier() else {
        eprintln!("collab: no notifier installed — put collab-notify.exe next to collab.exe");
        std::process::exit(1);
    };
    println!("using {}", h.display());
    let mut warned = false;
    raise(
        &h,
        "collab",
        &format!("#{} · test", config::channel()),
        "If you can see this, notifications work.",
        &config::channel(),
        &mut warned,
    );
    if !warned {
        println!("sent — it should be on screen now (and in the Action Centre)");
    }
}
